### PR TITLE
Use new format for errors

### PR DESCRIPTION
**Pull request by elgatovital**
Handles #842

https://docs.google.com/document/d/1foQ9DSMgO2iN-6u4JsJIDJWB7mUHX-D8ApZWvx6oKAo/edit#

Changes proposed in this pull request:
- Introduce new error classes
- Map exceptions to errors
- Show error messages using the new format

@gnosis/mobile-devs

### CODE REVIEW

File: app/src/main/java/io/gnosis/safe/ui/assets/coins/CoinsFragment.kt [hunk 1]
```diff
@@ -69,14 +68,8 @@ class CoinsFragment : BaseViewBindingFragment<FragmentCoinsBinding>() {
...
-                                when (action.error) {
-                                    is Offline -> {
-                                        snackbar(requireView(), R.string.error_no_internet)
-                                    }
-                                    else -> {
-                                        snackbar(requireView(), action.error.getErrorResForException())
-                                    }
-                                }
+                                val error = action.error.toError()
+                                snackbar(requireView(), error.message(requireContext(), R.string.error_description_assets_coins))
```
Comment (biafra23) [inline app/src/main/java/io/gnosis/safe/ui/assets/coins/CoinsFragment.kt:72]: The error messages are now so long it is hard to read them completely even with Snackbar.LENGTH_LONG. Could we go with LENGTH_INDEFINITE?

Comment (elgatovital) [reply app/src/main/java/io/gnosis/safe/ui/assets/coins/CoinsFragment.kt:72]: I would leave it like this for now and see what users will say. We can adjust it later.

Comment (biafra23) [reply app/src/main/java/io/gnosis/safe/ui/assets/coins/CoinsFragment.kt:72]: Have you tried it? Were you able to read the full message before it disappeared?
Setting this to 6000ms would make it disappear on its own. The user can swipe it to the right if he wants to get rid of it earlier.

Comment (elgatovital) [reply app/src/main/java/io/gnosis/safe/ui/assets/coins/CoinsFragment.kt:72]: let's do this as a separate task https://app.zenhub.com/workspaces/safe-multisig---mobile-5c1b98250e13551e8aae81eb/issues/gnosis/safe-android/1107

File: app/src/main/java/io/gnosis/safe/ErrorHandler.kt [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    ERROR_401(401, R.string.error_network_not_authorized_reason, R.string.error_network_not_authorized_fix),
+    ERROR_403(403, R.string.error_network_not_authorized_reason, R.string.error_network_not_authorized_fix),
+    ERROR_500(500, R.string.error_network_server_side_reason, R.string.error_network_server_side_fix),
+
+    // Common client errors
+    ERROR_1101(1101, R.string.error_client_safe_duplicate_reason, R.string.error_client_safe_duplicate_fix),
+    ERROR_1102(1102, R.string.error_client_address_invalid_reason, R.string.error_client_address_invalid_fix),
+    ERROR_1103(1103, R.string.error_client_seed_phrase_reason, R.string.error_client_seed_phrase_fix),
+    ERROR_1104(1104, R.string.error_client_tx_hash_validation_reason, R.string.error_client_tx_hash_validation_fix),
+    ERROR_1105(1105, R.string.error_client_unsupported_implementation_copy_reason, R.string.error_client_unsupported_implementation_copy_fix),
```
Comment (biafra23) [inline app/src/main/java/io/gnosis/safe/ErrorHandler.kt:49]: This is not used. Do we need it later, or should it be used already?

**Review by biafra23 — APPROVED** [review]
👍 
I really like the unification of al errors is one class and also having custom exception for everything. 
I would increase the time an error is shown, but that's not a blocker. Liliya will tell us if she doesn't approve :-)

File: app/src/main/java/io/gnosis/safe/ErrorHandler.kt (hunk 2 above)
Comment (elgatovital) [reply app/src/main/java/io/gnosis/safe/ErrorHandler.kt:49]: we might want to make safe validation more strict and allow adding safes that have min required version of the implementation copy.

Comment (elgatovital) [reply app/src/main/java/io/gnosis/safe/ErrorHandler.kt:49]: I'll add this to the next refinement. We have to adjust our safe validation

**Pull request merged**